Add support for CONST monomial to Bounds system

GiudGiud opened the pull request:
refs #23448

GiudGiud committed:
Add support for constant monomials for bounds system, refs #23448

GiudGiud committed:
Add test for bounded const monomial variable using constant BoundsAux, refs #23448

hugary1995 commented on framework/src/auxkernels/BoundsAuxBase.C:
```diff
@@ -33,23 +34,48 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
+  if (dummy.feType() != _var.feType())
+    paramError(
+        "variable",
+        "Dummy bound aux. variable and bounded variable must both finite element order and family");
```
```suggestion
        "Dummy bound aux. variable and bounded variable must use the same finite element order and family");
```

hugary1995 commented on framework/src/auxkernels/BoundsAuxBase.C:
```diff
@@ -33,23 +34,48 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
+  {
+    if (_current_node->n_dofs(_nl_sys.number(), _bounded_var.number()) > 0)
+    {
+      // The zero is for the component, this will only work for Lagrange variables
+      dof_id_type dof = _current_node->dof_number(_nl_sys.number(), _bounded_var.number(), 0);
+
+      Real bound = getBound();
+      _bounded_vector.set(dof, bound);
+    }
+  }
+  else
   {
-    // The zero is for the component, this will only work for Lagrange variables!
-    dof_id_type dof = _current_node->dof_number(_nl_sys.number(), _bounded_var.number(), 0);
+    if (_current_elem->n_dofs(_nl_sys.number(), _bounded_var.number()) > 0)
+    {
+      // The zero is for the component, this will only work for CONSTANT variables
+      dof_id_type dof = _current_elem->dof_number(_nl_sys.number(), _bounded_var.number(), 0);
 
-    Real bound = getBound();
-    _bounded_vector.set(dof, bound);
+      Real bound = getBound();
+      _bounded_vector.set(dof, bound);
+    }
```
Write a separate function which returns the current dof number of the bounded variable, then remove the code duplication here.

GiudGiud replied:
done. Only think was that this handled the "no local dof" case too. 

I think this matters if the block restriction is different between the dummy and the bounded

hugary1995 commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,142 @@
+  []
+[]
+
+[Executioner]
+  type = Steady
+
+  solve_type = 'PJFNK'
+[]
```
IIRC you have to use `-snes_type vinewtonrsls` or `-snes_type vinewtonssls` for the bounds to be active.

hugary1995 replied:
Can you also post of a plot of the result in this PR? Thanks.

GiudGiud replied:
It's passed in the tests spec. It s a good point though I ll move them

GiudGiud replied:
<img width="655" alt="Screen Shot 2023-02-16 at 3 36 47 PM" src="https://user-images.githubusercontent.com/25037151/219502702-fa5fb9c5-43b2-48dd-b356-6370e19e2afd.png">

hugary1995 reviewed (changes requested):
Looks good overall. Thanks again for taking care of this. Useful stuff.

GiudGiud committed:
Address review by Gary:
- getDofIndex independently of setting bounds
- explcitly specify in input file the solver method
- fix the error message

Update framework/src/auxkernels/BoundsAuxBase.C

Co-authored-by: Gary (Tianchen) Hu <thu@anl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1355225/) on 8943338 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23449/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23449/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ec30cd9096f4ea1d6ce8377698a190c06fc32dab`

GiudGiud commented:
Should be good now

lindsayad commented on framework/include/auxkernels/BoundsAuxBase.h:
```diff
@@ -49,4 +49,8 @@ class BoundsAuxBase : public AuxKernel
+
+private:
+  /// Return the current DOF index to apply the bound on
+  dof_id_type getDoFIndex();
```
```suggestion
  dof_id_type getDoFIndex() const;
```

lindsayad commented on framework/src/auxkernels/BoundsAuxBase.C:
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
+  if (!_var.isNodal() && (_var.feType().order != CONSTANT))
+    paramError("bounded_variable", "Bounded variable must be nodal or of a CONSTANT order!");
+
+  auto & dummy =
```
```suggestion
  const auto & dummy =
```

lindsayad commented on framework/src/auxkernels/BoundsAuxBase.C:
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
+  // Check that the dummy variable matches the bounded variable
+  if (dummy.feType() != _var.feType())
+    paramError("variable",
+               "Dummy bound aux. variable and bounded variable must use the same finite element "
```
```suggestion
               "Dummy bounds aux variable and bounded variable must use the same finite element "
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+[Mesh]
+  type = GeneratedMesh
+
+  dim = 1
+
+  xmin = 0
+  xmax = 1
+
+  nx = 10
```
```suggestion
  type = GeneratedMesh
  dim = 1
  xmin = 0
  xmax = 1
  nx = 10
```

lindsayad commented on framework/src/auxkernels/BoundsAuxBase.C:
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
 }
+
+dof_id_type
+BoundsAuxBase::getDoFIndex()
```
```suggestion
BoundsAuxBase::getDoFIndex() const
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+    order = CONSTANT
+    family = MONOMIAL
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+    type = Reaction
+    variable = u
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+    epsilon = -1
+    sigma = 6
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+    boundary = 1
+    value = 30
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+    epsilon = -1
+    sigma = 6
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+    bound_type = lower
+    bound_value = 0
+  []
+
```
```suggestion
```

lindsayad commented on test/tests/auxkernels/bounds/constant_bounds_elem.i:
```diff
@@ -0,0 +1,144 @@
+
+[Executioner]
+  type = Steady
+
```
```suggestion
```

hugary1995 commented on framework/src/auxkernels/BoundsAuxBase.C:
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
-    dof_id_type dof = _current_node->dof_number(_nl_sys.number(), _bounded_var.number(), 0);
+  dof_id_type dof = getDoFIndex();
 
+  if (dof != std::numeric_limits<dof_id_type>::max())
```
I think this condition will never be true based on the logic below. So maybe change this to an assertion.

GiudGiud replied:
I think it s a block restriction safeguard. This preserves the old behavior

lindsayad replied:
@hugary1995 it is possible to be true. Think of a first order Lagrange variable on a second order mesh. There will be no dofs on non-vertex nodes

hugary1995 replied:
I see. @GiudGiud can you add the reasoning as comments in the source code?

GiudGiud replied:
sure

lindsayad commented on test/tests/auxkernels/bounds/tests:
```diff
@@ -1,31 +1,57 @@
+    scale_refine = 3
+    design = 'ConstantBoundsAux.md'
+    issues = '#951 #14946'
+    requirement = "The system shall include the ability to compute an constant monomial field variable with a constant upper and lower bounds."
```
```suggestion
    requirement = "The system shall include the ability to compute a constant monomial field variable with constant upper and lower bounds."
```

lindsayad commented on test/tests/auxkernels/bounds/tests:
```diff
@@ -1,31 +1,57 @@
+  []
+
+  [exceptions]
+    requirement = "The system shall "
```
```suggestion
    requirement = "The system shall"
```
Details come out as bullets so space not necessary

lindsayad commented on test/tests/auxkernels/bounds/tests:
```diff
@@ -1,31 +1,57 @@
+      expect_out = "A variational inequalities solver must be used in conjunction with BoundsAux"
+      allow_warnings = true
+      cli_args = "Executioner/petsc_options_iname='' Executioner/petsc_options_value=''"
+      detail = 'emit a warning if a variational inequalities (VI) solver is not found when BoundsAux is provided.'
```
```suggestion
      detail = 'emit a warning if a variational inequalities (VI) solver is not found when the bounds system is used.'
```

GiudGiud committed:
Address Alex's review

- fix input file style
- more legible test requirements
- more const-ness

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1355451/) on 38978b8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23449/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23449/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ec30cd9096f4ea1d6ce8377698a190c06fc32dab`

GiudGiud commented:
should be good?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1355907/) on 946490f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23449/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1355921/) on 946490f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ec30cd</th>
      <th colspan=4>#23449 946490</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.72%</td>
      <td>84.72%</td>
      <td>+0.00%</td>
      <td>92.86%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>82898</td>
      <td>82907</td>
      <td>+9</td>
      <td>13</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14957</td>
      <td>14958</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23449/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23449/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23449/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
I think this PR has pretty strong endorsement at this point, down for a review

lindsayad commented:
Did someone circle a petition? 😆
